src/analysis/processing/qgsalgorithmpolygonstolines.cpp

```diff
@@ -0,0 +1,151 @@
+{
+  return QObject::tr( "Vector creation" );
+}
+
+QString QgsPolygonsToLinesAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeometry" );
+}
+
+QString QgsPolygonsToLinesAlgorithm::outputName() const
+{
+  return QObject::tr( "Lines" );
+}
+
+QString QgsPolygonsToLinesAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Converts polygons to lines" );
+}
+
+QString QgsPolygonsToLinesAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Converts polygons to lines." );
+}
+
+QgsPolygonsToLinesAlgorithm *QgsPolygonsToLinesAlgorithm::createInstance() const
+{
+  return new QgsPolygonsToLinesAlgorithm();
+}
+
+QList<int> QgsPolygonsToLinesAlgorithm::inputLayerTypes() const
+{
+  return QList< int >() << QgsProcessing::TypeVectorPolygon;
+}
+
+QgsFeatureList QgsPolygonsToLinesAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  Q_UNUSED( context )
+
+  QgsFeatureList result;
+  QgsFeature feat = feature;
+  if ( feat.hasGeometry() )
+    feat.setGeometry( convertToLines( feat.geometry() ) );
+
+  result << feat;
+  return result;
+}
+
+QgsGeometry QgsPolygonsToLinesAlgorithm::convertToLines( const QgsGeometry &geometry ) const
+{
+  auto rings = extractRings( geometry.constGet() );
+
+  QgsWkbTypes::Type resultType = outWkbType( geometry.wkbType() );
+
+  std::unique_ptr<QgsMultiCurve> lineGeometry;
+
+  if ( QgsWkbTypes::flatType( resultType ) == QgsWkbTypes::MultiLineString )
+    lineGeometry = qgis::make_unique<QgsMultiLineString>();
+  else
+    lineGeometry = qgis::make_unique<QgsMultiCurve>();
+
+  for ( auto ring : qgis::as_const( rings ) )
+    lineGeometry->addGeometry( ring );
+
+  return QgsGeometry( lineGeometry.release() );
+}
+
+QgsWkbTypes::Type QgsPolygonsToLinesAlgorithm::outWkbType( QgsWkbTypes::Type polygonWkbType ) const
+{
+  QgsWkbTypes::Type wkbType = QgsWkbTypes::NoGeometry;
+
+  if ( QgsWkbTypes::singleType( QgsWkbTypes::flatType( polygonWkbType ) ) == QgsWkbTypes::Polygon )
+    wkbType = QgsWkbTypes::MultiLineString;
+  else if ( QgsWkbTypes::singleType( QgsWkbTypes::flatType( polygonWkbType ) ) == QgsWkbTypes::CurvePolygon )
+    wkbType = QgsWkbTypes::MultiCurve;
+
+  if ( QgsWkbTypes::hasM( polygonWkbType ) )
+    wkbType = QgsWkbTypes::addM( wkbType );
+  if ( QgsWkbTypes::hasZ( polygonWkbType ) )
+    wkbType = QgsWkbTypes::addZ( wkbType );
+
+  return wkbType;
+}
+
+QList<QgsCurve *> QgsPolygonsToLinesAlgorithm::extractRings( const QgsAbstractGeometry *geom ) const
+{
+  QList<QgsCurve *> rings;
+
+  if ( QgsGeometryCollection *collection = qgsgeometry_cast<QgsGeometryCollection *>( geom ) )
+  {
+    QgsGeometryPartIterator parts = collection->parts();
+    while ( parts.hasNext() )
+      rings.append( extractRings( parts.next() ) );
+  }
+  else if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( geom ) )
+  {
+    rings.append( polygon->exteriorRing()->clone() );
```

Comment:
Do we need to guard against null exterior ring ? (i think we should).

Reply:
Also added a test for that